### ISSUE TITLE
Segmentation fault when importing bob.learn.em

### ISSUE THREAD

**Issue by 183amir** [bioidiap/bob.learn.em#17]
Backtrace:
```
gdb -ex r --args /opt/conda/envs/_test/bin/python -c "import bob.learn.em"
GNU gdb (GDB) Red Hat Enterprise Linux (7.2-83.el6)
Copyright (C) 2010 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>...
Reading symbols from /opt/conda/envs/_test/bin/python...done.
Starting program: /opt/conda/envs/_test/bin/python -c import\ bob.learn.em
warning: no loadable sections found in added symbol-file system-supplied DSO at 0x7ffff7ffa000
[Thread debugging using libthread_db enabled]
Detaching after fork from child process 73.
Missing separate debuginfo for /opt/conda/envs/_test/lib/python2.7/site-packages/numpy/core/../../../.././libgfortran.so.3
[New Thread 0x7ffff0176700 (LWP 75)]
[New Thread 0x7fffef975700 (LWP 76)]
[New Thread 0x7fffed174700 (LWP 77)]
warning: Source file is more recent than executable.

Program received signal SIGSEGV, Segmentation fault.
bob::extension::FunctionDoc::doc(unsigned int, unsigned int) const () at /opt/conda/envs/_build/lib/python2.7/site-packages/bob/extension/include/bob.extension/documentation.h:580
580	  if (description.empty()){
Missing separate debuginfos, use: debuginfo-install glibc-2.12-1.166.el6_7.7.x86_64
(gdb) bt
#0  bob::extension::FunctionDoc::doc(unsigned int, unsigned int) const () at /opt/conda/envs/_build/lib/python2.7/site-packages/bob/extension/include/bob.extension/documentation.h:580
#1  0x00007fffe441477b in _GLOBAL__sub_I_main.cpp () at bob/learn/em/main.cpp:18
#2  0x00007fffe4414826 in __do_global_ctors_aux () from /opt/conda/envs/_test/lib/python2.7/site-packages/bob/learn/em/_library.so
#3  0x00007fffe4391d03 in _init () from /opt/conda/envs/_test/lib/python2.7/site-packages/bob/learn/em/_library.so
#4  0x00007fff00000000 in ?? ()
#5  0x00007ffff7deb625 in _dl_init_internal () from /lib64/ld-linux-x86-64.so.2
#6  0x00007ffff7defe95 in dl_open_worker () from /lib64/ld-linux-x86-64.so.2
#7  0x00007ffff7deb286 in _dl_catch_error () from /lib64/ld-linux-x86-64.so.2
#8  0x00007ffff7def63a in _dl_open () from /lib64/ld-linux-x86-64.so.2
#9  0x00007ffff75d5f66 in dlopen_doit () from /lib64/libdl.so.2
#10 0x00007ffff7deb286 in _dl_catch_error () from /lib64/ld-linux-x86-64.so.2
#11 0x00007ffff75d629c in _dlerror_run () from /lib64/libdl.so.2
#12 0x00007ffff75d5ee1 in dlopen@@GLIBC_2.2.5 () from /lib64/libdl.so.2
#13 0x00007ffff7b25dde in _PyImport_GetDynLoadFunc (fqname=<value optimized out>, shortname=<value optimized out>, pathname=0xd34c60 "/opt/conda/envs/_test/lib/python2.7/site-packages/bob/learn/em/_library.so", fp=0xd39730) at Python/dynload_shlib.c:130
#14 0x00007ffff7b0a8d8 in _PyImport_LoadDynamicModule (name=0xcf9fb0 "bob.learn.em._library", pathname=0xd34c60 "/opt/conda/envs/_test/lib/python2.7/site-packages/bob/learn/em/_library.so", fp=0xd39730) at ./Python/importdl.c:42
#15 0x00007ffff7b08f81 in import_submodule (mod=0x7ffff7eaada8, subname=0xcf9fbd "_library", fullname=0xcf9fb0 "bob.learn.em._library") at Python/import.c:2704
#16 0x00007ffff7b091f4 in load_next (mod=0x7ffff7eaada8, altmod=0x7ffff7eaada8, p_name=<value optimized out>, buf=0xcf9fb0 "bob.learn.em._library", p_buflen=0x7fffffffe120) at Python/import.c:2519
#17 0x00007ffff7b09820 in import_module_level (name=<value optimized out>, globals=<value optimized out>, locals=<value optimized out>, fromlist=0x7ffff7ef69d0, level=<value optimized out>) at Python/import.c:2228
#18 PyImport_ImportModuleLevel (name=<value optimized out>, globals=<value optimized out>, locals=<value optimized out>, fromlist=0x7ffff7ef69d0, level=<value optimized out>) at Python/import.c:2292
#19 0x00007ffff7ae914f in builtin___import__ (self=<value optimized out>, args=<value optimized out>, kwds=<value optimized out>) at Python/bltinmodule.c:49
#20 0x00007ffff7a3fd23 in PyObject_Call (func=0x7ffff7fdffc8, arg=<value optimized out>, kw=<value optimized out>) at Objects/abstract.c:2546
#21 0x00007ffff7ae9633 in PyEval_CallObjectWithKeywords (func=0x7ffff7fdffc8, arg=0x7ffff7e935f0, kw=<value optimized out>) at Python/ceval.c:4219
#22 0x00007ffff7aee29e in PyEval_EvalFrameEx (f=<value optimized out>, throwflag=<value optimized out>) at Python/ceval.c:2622
#23 0x00007ffff7af3a2e in PyEval_EvalCodeEx (co=0x7ffff7eb08b0, globals=<value optimized out>, locals=<value optimized out>, args=<value optimized out>, argcount=0, kws=0x0, kwcount=0, defs=0x0, defcount=0, closure=0x0) at Python/ceval.c:3582
#24 0x00007ffff7af3b42 in PyEval_EvalCode (co=<value optimized out>, globals=<value optimized out>, locals=<value optimized out>) at Python/ceval.c:669
#25 0x00007ffff7b05a82 in PyImport_ExecCodeModuleEx (name=0x6b61b0 "bob.learn.em", co=0x7ffff7eb08b0, pathname=0x6b91e0 "/opt/conda/envs/_test/lib/python2.7/site-packages/bob/learn/em/__init__.pyc") at Python/import.c:713
#26 0x00007ffff7b081ce in load_source_module (name=0x6b61b0 "bob.learn.em", pathname=0x6b91e0 "/opt/conda/envs/_test/lib/python2.7/site-packages/bob/learn/em/__init__.pyc", fp=<value optimized out>) at Python/import.c:1103
#27 0x00007ffff7b08a2a in load_package (name=0x6b61b0 "bob.learn.em", pathname=<value optimized out>) at Python/import.c:1170
#28 0x00007ffff7b08f81 in import_submodule (mod=0x7ffff7eaad38, subname=0x6b61ba "em", fullname=0x6b61b0 "bob.learn.em") at Python/import.c:2704
#29 0x00007ffff7b091f4 in load_next (mod=0x7ffff7eaad38, altmod=0x7ffff7eaad38, p_name=<value optimized out>, buf=0x6b61b0 "bob.learn.em", p_buflen=0x7fffffffe710) at Python/import.c:2519
#30 0x00007ffff7b09860 in import_module_level (name=<value optimized out>, globals=<value optimized out>, locals=<value optimized out>, fromlist=0x7ffff7da6cd0, level=<value optimized out>) at Python/import.c:2236
#31 PyImport_ImportModuleLevel (name=<value optimized out>, globals=<value optimized out>, locals=<value optimized out>, fromlist=0x7ffff7da6cd0, level=<value optimized out>) at Python/import.c:2292
#32 0x00007ffff7ae914f in builtin___import__ (self=<value optimized out>, args=<value optimized out>, kwds=<value optimized out>) at Python/bltinmodule.c:49
#33 0x00007ffff7a3fd23 in PyObject_Call (func=0x7ffff7fdffc8, arg=<value optimized out>, kw=<value optimized out>) at Objects/abstract.c:2546
#34 0x00007ffff7ae9633 in PyEval_CallObjectWithKeywords (func=0x7ffff7fdffc8, arg=0x7ffff7ed15d0, kw=<value optimized out>) at Python/ceval.c:4219
#35 0x00007ffff7aee29e in PyEval_EvalFrameEx (f=<value optimized out>, throwflag=<value optimized out>) at Python/ceval.c:2622
#36 0x00007ffff7af3a2e in PyEval_EvalCodeEx (co=0x7ffff7edde30, globals=<value optimized out>, locals=<value optimized out>, args=<value optimized out>, argcount=0, kws=0x0, kwcount=0, defs=0x0, defcount=0, closure=0x0) at Python/ceval.c:3582
#37 0x00007ffff7af3b42 in PyEval_EvalCode (co=<value optimized out>, globals=<value optimized out>, locals=<value optimized out>) at Python/ceval.c:669
#38 0x00007ffff7b125cc in run_mod (str=0x601010 "import bob.learn.em\n", start=257, globals=0x7ffff7f81168, locals=0x7ffff7f81168, flags=<value optimized out>) at Python/pythonrun.c:1370
#39 PyRun_StringFlags (str=0x601010 "import bob.learn.em\n", start=257, globals=0x7ffff7f81168, locals=0x7ffff7f81168, flags=<value optimized out>) at Python/pythonrun.c:1333
#40 0x00007ffff7b138f0 in PyRun_SimpleStringFlags (command=0x601010 "import bob.learn.em\n", flags=0x7fffffffebc0) at Python/pythonrun.c:974
#41 0x00007ffff7b29457 in Py_Main (argc=<value optimized out>, argv=<value optimized out>) at Modules/main.c:589
#42 0x00007ffff6dd8d5d in __libc_start_main () from /lib64/libc.so.6
#43 0x0000000000400649 in _start ()
```
conda-forge pull request:
https://github.com/conda-forge/staged-recipes/pull/409

**Comment by 183amir** [comment]
The steps to reproduce this bug is: (you need `docker`)
```sh
git clone git@github.com:bioidiap/staged-recipes.git
cd staged-recipes
git checkout bob.learn.em
docker pull pelson/obvious-ci:latest_x64
./scripts/run_docker_build.sh
#now find the container that was created.
docker ps -l
docker start e0399f0838f0 #change this last parameter with the output of last command
docker attach e0399f0838f0
# now you are in bash in the docker
source activate _test #activate the test environment that just failed
# you should see a segfault trying to import this
python -c "import bob.learn.em"
yum install gdb python-debuginfo # install gdb
cp -R /opt/conda/envs/_test /opt/conda/envs/_build   # for gdb to find the header file
gdb -ex r --args /opt/conda/envs/_test/bin/python -c "import bob.learn.em"
```

**Comment by 183amir** [comment]
Just a note: Adding `BOB_SHORT_DOCSTRINGS=1` to environment variables did not fix this.

**Comment by siebenkopf** [comment]
The only thing that I might spot there is that the ``zt_norm`` functions:
https://github.com/bioidiap/bob.learn.em/blob/master/bob/learn/em/ztnorm.cpp#L18
are marked as member functions -- fourth parameter is true, see here:
https://github.com/bioidiap/bob.extension/blob/master/bob/extension/include/bob.extension/documentation.h#L90

However, I would be surprised if that would be the issue. It seems that the error occurs when accessing an internal string variable:
https://github.com/bioidiap/bob.extension/blob/master/bob/extension/include/bob.extension/documentation.h#L580
which is not modified in any other function than inside ``doc()``.

As the documentation functionality is used throughout all the bob packages, and it never generated any trouble like that, I would assume that someone else is messing around with the memory during the bindings, so that the internal representation of the ``description`` string is messed up.

**Comment by 183amir** [comment]
@siebenkopf I tried stepping through the program and when it reached https://github.com/bioidiap/bob.extension/blob/master/bob/extension/include/bob.extension/documentation.h#L580 `description` was not defined.

**Comment by siebenkopf** [comment]
What do you mean with not defined? It is a class member, so I don't see how it cannot be defined. Do you run the code in optimized mode? Or do you mean uninitialized?

**Comment by 183amir** [comment]
I don't remember the exact error and honestly I'm not good at gdb nor c++. Have you tried reproducing this bug and see it for yourself?

**Comment by siebenkopf** [comment]
Yes I tried, but I already got stuck in the ``docker pull pelson/obvious-ci:latest_x64`` command. Although I have installed ``docker`` on my system using the proposed ``sudo apt-get install docker``, when running the command, I get ``The program 'docker' is currently not installed.`` I am running that on Ubuntu 14.04. As I don't even know, what ``docker`` is used for, I gave up after that.

**Comment by 183amir** [comment]
you probably need to run docker with sudo

**Comment by siebenkopf** [comment]
I guess that the ``docker`` package in **Ubuntu** does not install the ``docker`` program. It simply does not exist, neither as me nor as root.

**Comment by 183amir** [comment]
Try visiting docker.com

**Comment by siebenkopf** [comment]
I hope that docker is not screwing around with my system. Why does it need sudo permissions to install some package? Does it install them system-wide?

**Comment by 183amir** [comment]
Docker is a virtualization software with much less overhead compared to virtual box. The docker pull command pulls cent os 6 image with conda already installed.

**Comment by siebenkopf** [comment]
Do you know how I can change the build parameters so that I get a debug build? In the current form, it is an optimized build, which I cannot debug nicely.

**Comment by 183amir** [comment]
You can set build parameters in your recipe like this:
```yaml
  script: CXXFLAGS="${CXXFLAGS} -DBOB_SHORT_DOCSTRINGS" python -B setup.py install --single-version-externally-managed --record record.txt
#or
  script: BOB_SHORT_DOCSTRINGS=1 python -B setup.py install --single-version-externally-managed --record record.txt
```
Which is a line in this file `staged-recipes/recipes/bob.learn.em/meta.yaml` in the git repo that you cloned.
Or you can remove that `script` line and add a `build.sh` file near `meta.yaml` and configure any build commands that you want. You can also look at http://conda.pydata.org/docs/building/build.html for more information.

**Comment by siebenkopf** [comment]
OK, I made it compile with:
```
script: CXXFLAGS="${CXXFLAGS} -ggdb -O0" CFLAGS="${CFLAGS} -ggdb -O0" python -B setup.py install --single-version-externally-managed --record record.txt
```
I am going to check if I can find anything now. However, I still don't understand why this is not working. It follows exactly the same pattern as for ``bob.ip.base``. However, when it tries to initialize the module definitions, it seems that all the variables are uninitialized, e.g., the function name is not even present in the documentation object:

```
(gdb) up
#1  0x00007fffd9e7898c in bob::extension::FunctionDoc::doc(unsigned int, unsigned int) const () at /opt/conda/envs/_build/lib/python2.7/site-packages/bob/extension/include/bob.extension/documentation.h:580
580	  if (description.empty()){

(gdb) up
#2  0x00007fffd9f10ba4 in __static_initialization_and_destruction_0 () at bob/learn/em/main.cpp:18
18	    zt_norm.doc()

(gdb) p zt_norm.function_name
$7 = {static npos = 18446744073709551615, _M_dataplus = {<std::allocator<char>> = {<__gnu_cxx::new_allocator<char>> = {<No data fields>}, <No data fields>}, _M_p = 0x0}}
```

So, apparently there is something wrong with the static initialization of the objects. Unfortunately, this is outside of my understanding. I can only assume that the compiler/linker did something wrong here (though this should not be the case).

Maybe, you can try to see if the same (or a similar) problem occurs when you comment out the functions inside the ``module_methods`` in the ``main.cpp``.

BTW: Have you tried to compile and run the tests in ``bob.ip.base``, or in any other library? Did they work?

**Comment by 183amir** [comment]
Several `bob` packages are already compiled and available in conda-forge: https://github.com/conda-forge?utf8=%E2%9C%93&query=bob 
I have not tried `bob.ip.base` yet because it depends on `vlfeat` and `vlfeat` is still not merged into `conda-forge`: https://github.com/conda-forge/staged-recipes/pull/362
Also you can also try building with buildout and test the code in docker.
```sh
git clone git@github.com:bioidiap/staged-recipes.git
cd staged-recipes
git checkout bob.learn.em
# checkout bob.learn.em source code
cd recipe/bob.learn.em
git clone git@github.com:bioidiap/bob.learn.em.git bob.learn.em-git
cd ../..
docker pull pelson/obvious-ci:latest_x64
./scripts/run_docker_build.sh
#now find the container that was created.
docker ps -l
docker start e0399f0838f0 #change this last parameter with the output of last command
docker attach e0399f0838f0
# now you are in bash in the docker
source activate _test #activate the test environment that just failed
# remove installed bob.learn.em
conda remove bob.learn.em
cd /recipe_root/bob.learn.em/bob.learn.em-git
python bootstrap-buildout.py
./bin/buildout
./bin/nosetests -sv
```

**Comment by siebenkopf** [comment]
OK, I tried that, but I am even unable to compile bob.learn.em. The compiler seems to be outdated, it does not recognize some command that build fine with all other compilers. To avoid downloading and compiling the bob packages inside bob.learn.em, I use some command line parameters that will use the pre-installed dependent bob packages.
Anyways, I get the error message:

```
BOB_BUILD_PARALLEL=8 nice -n15 bin/buildout buildout:debug=false buildout:develop=. buildout:extensions=bob.buildout buildout:auto-checkout=
Develop: '/conda-recipes/bob.learn.em/bob.learn.em-git/.'
bob.buildout.tools: taking requirement `setuptools' (20.10.1) from `/conda-recipes/bob.learn.em/bob.learn.em-git/eggs/setuptools-20.10.1-py2.7.egg'
bob.buildout.tools: taking requirement `bob.extension' (2.0.11) from `/opt/conda/envs/_test/lib/python2.7/site-packages'
bob.buildout.tools: taking requirement `bob.blitz' (2.0.8) from `/opt/conda/envs/_test/lib/python2.7/site-packages'
bob.buildout.tools: taking requirement `bob.core' (2.1.2) from `/opt/conda/envs/_test/lib/python2.7/site-packages'
bob.buildout.tools: taking requirement `bob.io.base' (2.0.8) from `/opt/conda/envs/_test/lib/python2.7/site-packages'
bob.buildout.tools: taking requirement `bob.sp' (2.0.4) from `/opt/conda/envs/_test/lib/python2.7/site-packages'
bob.buildout.tools: taking requirement `bob.math' (2.0.3) from `/opt/conda/envs/_test/lib/python2.7/site-packages'
bob.buildout.tools: taking requirement `bob.learn.activation' (2.0.4) from `/opt/conda/envs/_test/lib/python2.7/site-packages'
bob.buildout.tools: taking requirement `bob.learn.linear' (2.0.7) from `/opt/conda/envs/_test/lib/python2.7/site-packages'
...
[ 58%] Building CXX object CMakeFiles/bob_learn_em.dir/conda-recipes/bob.learn.em/bob.learn.em-git/bob/learn/em/cpp/FABaseTrainer.cpp.o
/usr/bin/c++   -DBLITZ_VERSION=\"0.10\" -DHAVE_BLITZ=1 -DNO_IMPORT_ARRAY=1 -DNPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION -DPY_ARRAY_UNIQUE_SYMBOL=BOB_NUMPY_C_API -Dbob_learn_em_EXPORTS -I/opt/conda/envs/_test/lib/python2.7/site-packages/bob/learn/activation/include -I/opt/conda/envs/_test/lib/python2.7/site-packages/bob/sp/include -I/opt/conda/envs/_test/lib/python2.7/site-packages/bob/math/include -I/opt/conda/envs/_test/lib/python2.7/site-packages/bob/learn/linear/include -I/opt/conda/envs/_test/lib/python2.7/site-packages/bob/core/include -I/opt/conda/envs/_test/lib/python2.7/site-packages/bob/io/base/include -isystem /opt/conda/envs/_test/include -I/conda-recipes/bob.learn.em/bob.learn.em-git/bob/learn/em/include -I/opt/conda/envs/_test/lib/python2.7/site-packages/bob/blitz/include -isystem /opt/conda/envs/_test/lib/python2.7/site-packages/numpy/core/include  -std=c++0x -pedantic -Wall -pthread -O0 -g -DBOB_DEBUG -DBZ_DEBUG  -fPIC   -o CMakeFiles/bob_learn_em.dir/conda-recipes/bob.learn.em/bob.learn.em-git/bob/learn/em/cpp/FABaseTrainer.cpp.o -c /conda-recipes/bob.learn.em/bob.learn.em-git/bob/learn/em/cpp/FABaseTrainer.cpp
In file included from /opt/conda/envs/_test/lib/python2.7/site-packages/bob/core/include/bob.core/array_random.h:12,
                 from /conda-recipes/bob.learn.em/bob.learn.em-git/bob/learn/em/cpp/FABaseTrainer.cpp:14:
/opt/conda/envs/_test/lib/python2.7/site-packages/bob/core/include/bob.core/random.h:19: error: expected unqualified-id before ‘using’
/opt/conda/envs/_test/lib/python2.7/site-packages/bob/core/include/bob.core/random.h:22: error: expected unqualified-id before ‘using’
/opt/conda/envs/_test/lib/python2.7/site-packages/bob/core/include/bob.core/random.h:25: error: expected unqualified-id before ‘using’
/opt/conda/envs/_test/lib/python2.7/site-packages/bob/core/include/bob.core/random.h:28: error: expected unqualified-id before ‘using’
/opt/conda/envs/_test/lib/python2.7/site-packages/bob/core/include/bob.core/random.h:31: error: expected unqualified-id before ‘using’
In file included from /conda-recipes/bob.learn.em/bob.learn.em-git/bob/learn/em/cpp/FABaseTrainer.cpp:14:
/opt/conda/envs/_test/lib/python2.7/site-packages/bob/core/include/bob.core/array_random.h: In function ‘void bob::core::array::randn(boost::random::mt19937&, blitz::Array<T, N>&, T, T)’:
/opt/conda/envs/_test/lib/python2.7/site-packages/bob/core/include/bob.core/array_random.h:32: error: ‘normal_distribution’ is not a member of ‘bob::core::random’
/opt/conda/envs/_test/lib/python2.7/site-packages/bob/core/include/bob.core/array_random.h:32: error: expected primary-expression before ‘>’ token
/opt/conda/envs/_test/lib/python2.7/site-packages/bob/core/include/bob.core/array_random.h:35: error: there are no arguments to ‘distrib’ that depend on a template parameter, so a declaration of ‘distrib’ must be available
/opt/conda/envs/_test/lib/python2.7/site-packages/bob/core/include/bob.core/array_random.h:35: note: (if you use ‘-fpermissive’, G++ will accept your code, but allowing the use of an undeclared name is deprecated)
```

Apparently, the compiler does not know, what to do with the ``using`` keyword (a C++11 feature), that all other compilers work well with.

BTW: in my case the directories are ``cd recipes/bob.learn.em`` and ``cd /conda-recipes/bob.learn.em/bob.learn.em-git``.

**Comment by 183amir** [comment]
Sorry you need to run `conda install pkg-config cmake gcc` after you do `conda remove bob.learn.em`.
The instructions I have written in my comment are from memory so your directories should be correct.

**Comment by siebenkopf** [comment]
I could figure out cmake on my own, but not gcc :-)

Unfortunately, I don't get the source code any more, but I see only:
```
gdb -ex r --args /opt/conda/envs/_test/bin/python -c "import bob.learn.em"
[...]
Program received signal SIGSEGV, Segmentation fault.
0x00007fffda1695cf in bob::extension::FunctionDoc::doc(unsigned int, unsigned int) const () from bob/learn/em/_library.so
Missing separate debuginfos, use: debuginfo-install glibc-2.12-1.166.el6_7.7.x86_64
```

The trick, i.e., to ``cp -R /opt/conda/envs/_test /opt/conda/envs/_build`` doesn't help.

**Comment by 183amir** [comment]
Did you run `yum install gdb python-debuginfo # install gdb`?

**Comment by 183amir** [comment]
@siebenkopf I updated the instructions in my first comment. Also with the help of Andre we realized that this could be a [static-initialization-order-fiasco](https://stackoverflow.com/questions/3035422/static-initialization-order-fiasco). When the code reaches `bob/learn/em/main.cpp:13`, `zt_norm` is not initialized properly. I currently do not know how to prevent this but I am working on it.

**Comment by siebenkopf** [comment]
In fact, I had installed ``yum install -y gdb python-debuginfo``, but only after I compiled. I recompiled everything, and I now get the error, and I am able to debug. I will let you know if I can find out something.

**Comment by siebenkopf** [comment]
So, I have found a solution, but I didn't get the reason why it failed in the first place, and only in the given scenario. I think that there is something wrong with the static initialization of the variables:

1. The variables are declared ``extern`` in the ``main.h``: 
https://github.com/bioidiap/bob.learn.em/blob/master/bob/learn/em/main.h#L270

2. The variables are implemented in the ``ztnorm.cpp``:
https://github.com/bioidiap/bob.learn.em/blob/master/bob/learn/em/ztnorm.cpp#L13

3. The variables are used in the ``main.cpp``:
https://github.com/bioidiap/bob.learn.em/blob/master/bob/learn/em/main.cpp#L13

For some reason, this does not seem to work, when I debug the code, the ``zt_norm`` (and other) variables are not initialized, i.e., all internal pointers point to ``0x0`` and the strings are empty. Hence, I assume that the initialization in the ``ztnorm.cpp`` did not happen yet, which is weired as these objects are global and inside the ``_library.so`` (I have checked ``readelf -s bob/learn/em/_library.so`` and ``zt_norm`` appears as a GLOBAL object). However, these objects are not picked up correctly.

Now, the solution (which I don't really like) is to **move** the declaration of the documentations **into** the ``main.cpp``, before
https://github.com/bioidiap/bob.learn.em/blob/master/bob/learn/em/main.cpp#L13
Apparently, now the variables are picked up correctly, and everything works. Interestingly, when used inside the actual function binding in 
https://github.com/bioidiap/bob.learn.em/blob/master/bob/learn/em/ztnorm.cpp#L29
the variable is picked up correctly, and the correct parameter list is used (I assume, the tests pass).

The reason, why I don't like the solution is that now, the documentation of a function is decoupled from their implementation (i.e., the binding). And, once again, I don't understand, why the global objects are not picked up correctly. It might be some bug inside conda, or docker, or gcc, or ..., which appears only in this combination. Note that the code runs fine since month on any other tested platform.

**Comment by anjos** [comment]
My 2 cents on this: I think most documentation strings are marked as `static` on other packages, whereas in this one they are not. Static initialisation is a pain - we must ensure all variables required for static work are available before we start it. In this case, the "work" is the module declaration, so the documentation strings must be ready when the module initialises. The only way to guarantee this is to make the documentation strings also static. Have you tried that solution? This would be cleaner than moving the documentation strings into `main.cpp`. This is also the solution used in other packages. For example:

https://github.com/bioidiap/bob.learn.linear/blob/master/bob/learn/linear/lda.cpp#L21

What do you think?

**Comment by 183amir** [comment]
The problem is these documentations are used in another file and defining
them static would make them private. My suggestion is to write functions
that would return these static variables. This way, the implementation
would be in their own file and static but also can be used outside of the
files.

**Comment by 183amir** [comment]
Ok guys, I managed to fix with including ztnorm.cpp in main.cpp. What do you think?
Here is my patch: 
```patch
diff --git bob/learn/em/linear_scoring.cpp bob/learn/em/linear_scoring.cpp
index c475ff0..e5b592d 100644
--- bob/learn/em/linear_scoring.cpp
+++ bob/learn/em/linear_scoring.cpp
@@ -69,7 +69,7 @@ static inline bool f(PyObject* o){return o != 0 && PyObject_IsTrue(o) > 0;}
 
 
 /*** linear_scoring ***/
-bob::extension::FunctionDoc linear_scoring1 = bob::extension::FunctionDoc(
+static auto linear_scoring1 = bob::extension::FunctionDoc(
   "linear_scoring",
   "",
   0,
@@ -84,7 +84,7 @@ bob::extension::FunctionDoc linear_scoring1 = bob::extension::FunctionDoc(
 .add_return("output","array_like<float,1>","Score");
 
 
-bob::extension::FunctionDoc linear_scoring2 = bob::extension::FunctionDoc(
+static auto linear_scoring2 = bob::extension::FunctionDoc(
   "linear_scoring",
   "",
   0,
@@ -101,7 +101,7 @@ bob::extension::FunctionDoc linear_scoring2 = bob::extension::FunctionDoc(
 
 
 
-bob::extension::FunctionDoc linear_scoring3 = bob::extension::FunctionDoc(
+static auto linear_scoring3 = bob::extension::FunctionDoc(
   "linear_scoring",
   "",
   0,
diff --git a/bob/learn/em/main.cpp b/bob/learn/em/main.cpp
index dc413e8..49e4c3a 100644
--- bob/learn/em/main.cpp
+++ bob/learn/em/main.cpp
@@ -10,6 +10,10 @@
 #endif
 #include "main.h"
 
+// since their variables are static and need to be used here.
+#include "ztnorm.cpp"
+#include "linear_scoring.cpp"
+
 static PyMethodDef module_methods[] = {
   {
     zt_norm.name(),
diff --git a/bob/learn/em/main.h b/bob/learn/em/main.h
index 1d61e87..f22cae1 100644
--- bob/learn/em/main.h
+++ bob/learn/em/main.h
@@ -265,21 +265,21 @@ bool init_BobLearnEMEMPCATrainer(PyObject* module);
 int PyBobLearnEMEMPCATrainer_Check(PyObject* o);
 
 
-//ZT Normalization
-PyObject* PyBobLearnEM_ztNorm(PyObject*, PyObject* args, PyObject* kwargs);
-extern bob::extension::FunctionDoc zt_norm;
+// //ZT Normalization
+// PyObject* PyBobLearnEM_ztNorm(PyObject*, PyObject* args, PyObject* kwargs);
+// extern bob::extension::FunctionDoc zt_norm;
 
-PyObject* PyBobLearnEM_tNorm(PyObject*, PyObject* args, PyObject* kwargs);
-extern bob::extension::FunctionDoc t_norm;
+// PyObject* PyBobLearnEM_tNorm(PyObject*, PyObject* args, PyObject* kwargs);
+// extern bob::extension::FunctionDoc t_norm;
 
-PyObject* PyBobLearnEM_zNorm(PyObject*, PyObject* args, PyObject* kwargs);
-extern bob::extension::FunctionDoc z_norm;
+// PyObject* PyBobLearnEM_zNorm(PyObject*, PyObject* args, PyObject* kwargs);
+// extern bob::extension::FunctionDoc z_norm;
 
 
-//Linear scoring
-PyObject* PyBobLearnEM_linear_scoring(PyObject*, PyObject* args, PyObject* kwargs);
-extern bob::extension::FunctionDoc linear_scoring1;
-extern bob::extension::FunctionDoc linear_scoring2;
-extern bob::extension::FunctionDoc linear_scoring3;
+// //Linear scoring
+// PyObject* PyBobLearnEM_linear_scoring(PyObject*, PyObject* args, PyObject* kwargs);
+// extern bob::extension::FunctionDoc linear_scoring1;
+// extern bob::extension::FunctionDoc linear_scoring2;
+// extern bob::extension::FunctionDoc linear_scoring3;
 
 #endif // BOB_LEARN_EM_MAIN_H
diff --git a/bob/learn/em/ztnorm.cpp b/bob/learn/em/ztnorm.cpp
index adc8a6f..3fce1ab 100644
--- bob/learn/em/ztnorm.cpp
+++ bob/learn/em/ztnorm.cpp
@@ -10,7 +10,7 @@
 #include "main.h"
 
 /*** zt_norm ***/
-bob::extension::FunctionDoc zt_norm = bob::extension::FunctionDoc(
+static auto zt_norm = bob::extension::FunctionDoc(
   "ztnorm",
   "Normalise raw scores with ZT-Norm."
   "Assume that znorm and tnorm have no common subject id.",
@@ -70,7 +70,7 @@ PyObject* PyBobLearnEM_ztNorm(PyObject*, PyObject* args, PyObject* kwargs) {
 
 
 /*** t_norm ***/
-bob::extension::FunctionDoc t_norm = bob::extension::FunctionDoc(
+static auto t_norm = bob::extension::FunctionDoc(
   "tnorm",
   "Normalise raw scores with T-Norm",
   0,
@@ -107,7 +107,7 @@ PyObject* PyBobLearnEM_tNorm(PyObject*, PyObject* args, PyObject* kwargs) {
 
 
 /*** z_norm ***/
-bob::extension::FunctionDoc z_norm = bob::extension::FunctionDoc(
+static auto z_norm = bob::extension::FunctionDoc(
   "znorm",
   "Normalise raw scores with Z-Norm",
   0,
diff --git a/setup.py b/setup.py
index 3f56083..5a0573c 100644
--- setup.py
+++ setup.py
@@ -114,9 +114,9 @@ setup(
 
           "bob/learn/em/plda_trainer.cpp",
 
-          "bob/learn/em/ztnorm.cpp",
-
-          "bob/learn/em/linear_scoring.cpp",
+          # include the module methods directly in main.cpp
+          # "bob/learn/em/ztnorm.cpp",
+          # "bob/learn/em/linear_scoring.cpp",
 
           "bob/learn/em/main.cpp",
         ],
```

**Comment by 183amir** [comment]
@anjos Looking at here: https://github.com/bioidiap/bob.measure/blob/master/bob/measure/main.cpp all the bindings and documentation for module functions are written in one file.

**Comment by siebenkopf** [comment]
@anjos: Please do not get fooled by the ``static`` keyword: http://thesmithfam.org/blog/2010/06/12/the-many-meanings-of-the-c-static-keyword/
All global variables by default have ``static`` linkage, independent of whether you declare them ``static`` or not. For global variables (as well as for global functions), declaring them ``static`` just marks them not to be exported to other files, and they do not appear in the shared object. In our case, as @183amir pointed out, this wouldn't work, as we define the variables in another file as we use them. For the documentation strings, this might be a good idea, but it is no solution to our problem.

When I read the above post in more detail, it is written there:
> [...] the language makes no guarantee about the order in which static variables are initialized. In my experience, static variables in the same file are initialized in a “sane” order [...]
Hence, I guess this is where our issue comes from.
Anyways, it seems the standard does not even define that static variables **in the same file** are initialized in order. 

@183amir: I don't like to include all module methods in the ``main.cpp``. For ``bob.learn.em`` this might be feasible, but, e.g., for ``bob.ip.base`` there are tons of module-level functions, which would blow up the ``main.cpp`` dramatically.
I could go with the idea of @183amir, to have functions that return the function-local static documentation strings. According to the link above:
> Static function-local variables are initialized the first time the function is called [...]
This might do the trick, and this would also avoid to have the documentation exported in the shared object (although we would now export the function to generate them).

**Comment by 183amir** [comment]
conda-forge introduced a new gcc and now this is not an issue anymore: https://circleci.com/gh/conda-forge/staged-recipes/2620

**Comment by 183amir** [comment]
Since this is fixed with a different `gcc`, I will close this.
